### PR TITLE
Add U-net style skip connections to the semantic-segmentation example

### PR DESCRIPTION
**Pull request by reunanen**
To address the [recent interest](https://github.com/davisking/dlib/issues/1236#issuecomment-447581865) in a U-net implementation, this PR adds a possibly-useful `resize_to_prev` layer, and upgrades the semantic-segmentation example.

Compared to the corresponding network without skip connections (see #943), the updated example should be able to draw more details:

image

However, overall it's not too much more accurate on the PASCAL VOC 2012 data -- this time around, the training should finish like this (note the lower loss but not much better accuracy, when compared to numbers reported in #943):

```
step#: 117901  learning rate: 0.001  average loss: 0.0239372    steps without apparent progress: 4941
step#: 118018  learning rate: 0.001  average loss: 0.0238844    steps without apparent progress: 4878
step#: 118135  learning rate: 0.001  average loss: 0.0235096    steps without apparent progress: 4840
step#: 118252  learning rate: 0.0001  average loss: 0.0232595    steps without apparent progress: 4178
step#: 118369  learning rate: 0.0001  average loss: 0.0243033    steps without apparent progress: 4282
step#: 118486  learning rate: 0.0001  average loss: 0.0214507    steps without apparent progress: 4412
step#: 118603  learning rate: 0.0001  average loss: 0.0246106    steps without apparent progress: 4516
step#: 118720  learning rate: 0.0001  average loss: 0.0239061    steps without apparent progress: 4646
step#: 118837  learning rate: 0.0001  average loss: 0.0219395    steps without apparent progress: 4750
step#: 118954  learning rate: 0.0001  average loss: 0.0220728    steps without apparent progress: 4880
Saved state to pascal_voc2012_trainer_state_file.dat_
step#: 119063  learning rate: 0.0001  average loss: 0.0224262    steps without apparent progress: 4903
Saved state to pascal_voc2012_trainer_state_file.dat
saving network
Testing the network...
train accuracy  :  0.991385
val accuracy    :  0.851742
```

However, on some other datasets the U-net approach can be very useful. (I guess in particular when there are lots of details to draw, but the classification task itself is not too difficult.)

~Note that I still want to find a network architecture that is clean, performs well, and compiles in finite time using MSVC++ (see the comment below). So I'd prefer not to have this merged just yet. But at the same time wanted to make it available already, in case e.g. @goldbattle wants to try it out.~

### CODE REVIEW

File: examples/dnn_semantic_segmentation_ex.h [hunk 1]
```diff
@@ -145,55 +145,129 @@ template <int N, typename SUBNET> using ares_up   = dlib::relu<residual_up<block
...
-template <typename SUBNET> using level4t = dlib::repeat<2,res64,res_up<64,SUBNET>>;
+// ----------------------------------------------------------------------------------------
 
-template <typename SUBNET> using alevel1t = dlib::repeat<2,ares512,ares_up<512,SUBNET>>;
-template <typename SUBNET> using alevel2t = dlib::repeat<2,ares256,ares_up<256,SUBNET>>;
-template <typename SUBNET> using alevel3t = dlib::repeat<2,ares128,ares_up<128,SUBNET>>;
-template <typename SUBNET> using alevel4t = dlib::repeat<2,ares64,ares_up<64,SUBNET>>;
+// Microsoft Visual C++ compilers choke when trying to build networks that are too deep.
+// So let's limit the depth a little. On the PASCAL VOC 2012 data, the accuracy hit is
+// not even very significant.
```
Comment (reunanen) [inline examples/dnn_semantic_segmentation_ex.h]: I could not quite decide if I should keep the example small enough that MSVC can build it, or have it be closer to the full U-net implementation with all the skip connections.

So at least for now, I decided to have both (using `#ifdef`s), even if the resulting code is a bit of a mess. But for simplicity, am willing to go for just one implementation.

(After this has been decided, I'll make the trained `v2` model(s) available.)

Note that on the PASCAL VOC 2012 data, there's not too much difference in terms of accuracy.

File: examples/dnn_semantic_segmentation_ex.h [hunk 2]
```diff
@@ -145,55 +145,129 @@ template <int N, typename SUBNET> using ares_up   = dlib::relu<residual_up<block
...
-using net_type = dlib::loss_multiclass_log_per_pixel<
-                            dlib::cont<class_count,7,7,2,2,
-                            level4t<level3t<level2t<level1t<
-                            level1<level2<level3<level4<
-                            dlib::max_pool<3,3,2,2,dlib::relu<dlib::bn_con<dlib::con<64,7,7,2,2,
-                            dlib::input<dlib::matrix<dlib::rgb_pixel>>
-                            >>>>>>>>>>>>>>;
+using bnet_type = dlib::loss_multiclass_log_per_pixel<
+                              dlib::cont<class_count,1,1,1,1,
+#if LIMIT_LAYER_COUNT_BECAUSE_OF_COMPILER_ISSUES == 0
```
Comment (davisking) [inline examples/dnn_semantic_segmentation_ex.h]: It's a little easier to read if written `#ifdef LIMIT_LAYER_COUNT_BECAUSE_OF_COMPILER_ISSUES`

**Comment by davisking** [comment]
Sweet, thanks for the PR.  I'll look it over in detail in a little bit. 

But about MSVC, I thought deeper networks were working in the newest version now that we have a few workarounds in there.  I guess this one is still just too deep?  Are you using the newest MSVC?

File: examples/dnn_semantic_segmentation_ex.h (hunk 2 above)
Comment (reunanen) [reply examples/dnn_semantic_segmentation_ex.h]: Or actually `#ifndef LIMIT_LAYER_COUNT_BECAUSE_OF_COMPILER_ISSUES` -- but as I wrote, I'd like to get rid of that anyway.

**Comment by reunanen** [comment]
@davisking Ok, that could be. I'm still mainly on Visual Studio 2015. Will try 2017 as well.

**Comment by reunanen** [comment]
> But about MSVC, I thought deeper networks were working in the newest version now that we have a few workarounds in there. I guess this one is still just too deep? Are you using the newest MSVC?

Woohoo, the latest MSVC (19.16.27025.1) _does_ indeed fix essentially all my speed issues! This will _change my life_! Thanks so much for pointing this out, Davis! 🚀 

... so, will be back with an updated U-net structure shortly.

**Comment by davisking** [comment]
Sweet :)

**Comment by reunanen** [comment]
Here's a trained network: https://www.dropbox.com/s/etdk5nlxa22e4tf/semantic_segmentation_voc2012net_v2.dnn

**Comment by davisking** [comment]
Sweet, so is this ready to go? :)

As an aside, the semantic segmentation training and eval programs both contain
```
        // Crop the net output to be exactly the same size as the input.
        const chip_details chip_details(
            centered_rect(temp.nc() / 2, temp.nr() / 2, input_image.nc(), input_image.nr()),
            chip_dims(input_image.nr(), input_image.nc())
        );
        extract_image_chip(temp, chip_details, net_output, interpolate_nearest_neighbor());
```
Couldn't that be done more straightforwardly using this? 
```
temp.set_size(net_output.nr(), net_output.nc());
resize_image(net_output,temp,interpolate_nearest_neighbor());
```

**Comment by reunanen** [comment]
> Sweet, so is this ready to go? :)

Naming issues aside, I think it is.

If the layer name is [changed](https://github.com/davisking/dlib/pull/1600#discussion_r245481852), then the [pre-trained net](https://github.com/davisking/dlib/pull/1600#issuecomment-451663677) will not work as it is, right? (Deserialization will fail?)

That's not a problem, I can re-train it. (I guess re-training is better than trying to edit the binary file...)

> Couldn't that be done more straightforwardly using this?

I guess the idea is that the `extract_image_chip` variety crops (un-pads) the image, and does not really rescale it.

**Comment by davisking** [comment]
Cool.  Renaming sounds good. Don’t retrain though. Just edit serialize and not deserialize. Then load and save the file. Way easier. 

If it’s really just dropping rows and columns off the border then using sub_image is a more straightforward way. In particular, it’s not super obvious to me that extract_image_chips isn’t doing some interpolation.

**Comment by reunanen** [comment]
It's not obvious to me either. But I guess `sub_image` will not pad, if necessary? So `extract_image_chip` looks a little safer to me.

File: dlib/dnn/layers_abstract.h [hunk 3]
```diff
@@ -2382,6 +2382,56 @@ namespace dlib
...
+                defined above.  This layer resizes the output channels of the tagged layer to
+                have the same number of rows and columns as the output of the previous layer.
+
+                This layer uses bilinear interpolation. If the sizes match already, then it
+                simply copies the data.
+
+                Therefore, you supply a tag via resize_to_prev's template argument that tells
+                it what layer to use for the target size.
+
+                If tensor A is resized to size of tensor B, then a tensor C is produced such
```
[the remark this answers is not in the document]
Comment (reunanen) [reply dlib/dnn/layers_abstract.h]: Fixed.

**Comment by reunanen** [comment]
> Don’t retrain though. Just edit serialize and not deserialize. Then load and save the file.

This is now done. Same url as previously: https://www.dropbox.com/s/etdk5nlxa22e4tf/semantic_segmentation_voc2012net_v2.dnn

**Comment by davisking** [comment]
Sweet, looks good.  Thanks for another nice PR :)

I uploaded the model to dlib.net. So it's all ready to go.  Also feel free to make a PR at https://github.com/davisking/dlib-models for the model if you want.  I'm sort of indifferent about putting models there.  I only update it when some user bothers me about it :).  But you can put the model there as well if you feel so inclined.

**Pull request merged**